### ISSUE TITLE
EMA trains faster

### ISSUE THREAD

**Issue by pfriesch** [mkotha/WaveRNN#1]
Hi,

Cool project!

When I was trying VQVAE I found that using a moving average like described in the appendix trained a lot faster and gave better results! There is a [zalandoresearch](https://github.com/zalandoresearch/pytorch-vq-vae) repo that has a open source example. It is a bit hard to parallelize though, since it does not depend on the optimizer to learn the embedding.

Cheers 👍

**Comment by mkotha** [comment]
Thanks for the information. It sounds like a good modification to try!

**Comment by mkotha** [comment]
Let's leave this open - someone (perhaps I) might want to implement this.

**Comment by mazzzystar** [comment]
I've tried on WaveNet+VQVAE with EMA, and it seems that using EMA only on`VQ` space can get a reasonable result.

Here is my [result.zip](https://github.com/mkotha/WaveRNN/files/2976179/demo.zip), not as good as the official demo, though.


The implementation is just follow https://discuss.pytorch.org/t/how-to-apply-exponential-moving-average-decay-for-variables/10856/3 , glad if you find it helps.

**Comment by pfriesch** [comment]
If I am not mistaken, the linked example is a regularization scheme for the model parameters?! That is not the EMA to learn just the embeddings? Do you have an example of your implementation?

**Comment by mazzzystar** [comment]
What do you mean by `the linked example is a regularization scheme for the model parameters`?

I mean it's the final audio result of my implementation of `VQ-VAE` with `EMA` only on embedding.

**Comment by pfriesch** [comment]
Well, the forum post you linked is not the EMA that is described in the VQVAE paper. Did you modify the implementation to follow the VQVAE paper?

**Comment by mazzzystar** [comment]
Actually I think they also use `EMA` only on embedding.

You can see the implementation of [sonnet/vqvae.py](https://github.com/deepmind/sonnet/blob/c2b3edaca63032985897c6e629a55cd02c19d53e/sonnet/python/modules/nets/vqvae.py#L124)， I don't really read their implementation of `EMA`.

> The difference between VectorQuantizerEMA and VectorQuantizer is that
  this module uses exponential moving averages to update the embedding vectors
  instead of an auxiliary loss.

**Comment by pfriesch** [comment]
I mean it is called both exponential moving average (EMA). The forum post which you posted and what they do in the VQVAE paper (or in the sonnet code) are fundamentally different things. Even if they are only applied to the embedding parameters.  The forum post seems to be a regularization strategy (I am not sure what it aims to do exactly), while the VQVAE EMA is a different learning algorithm for the embeddings. 

I am not sure if you tried the correct VQVAE EMA learning algorithm, since you wrote that you followed to forum post.

**Comment by mazzzystar** [comment]
Yes I follow the forum method of exponential moving average. Sorry I didn't really realize they are different things. To my knowledge `EMA` is a regularization strategy, so I did not check the `EMA` idea of `VQVAE`. Thanks for your reminder, I'll check the paper tomorrow.

**Comment by pfriesch** [comment]
Cool 👍  Looking forward to the samples 😄

**Comment by mazzzystar** [comment]
I quickly scan the Appendix of the VQVAE paper.

It seems they use the method as below:
Remove `sg(z_ex), e) ** 2` from  loss function below
```
L = loss_recon + sg(z_ex), e) ** 2 + 0.25 * (sg(z_qx), e))
```

And update the embedding `e` as below:
* Suppose currently the embedding center is `e_i`, there are `N` encoder outputs closet to it. 
* Then `e_i` will update itself with EMA.

What's your idea on that ?

**Comment by pfriesch** [comment]
Yes, sounds correct. You can look at [this notebook](https://github.com/zalandoresearch/pytorch-vq-vae/blob/master/vq-vae.ipynb) for a working example.
